code/Architecture.md

```diff
@@ -0,0 +1,47 @@
+# Architecture
+
+This document describes how things are implemented.
+
+## Layers
+
+The code-level architecture is categorized into a few layers:
+
+* [Peer](code/Peer.md) - is a top-level concept that wires all dependencies together for Satellite, Storage Node or Uplink;
+* Subsystem - is a collection of endpoints and services related to a single problem;
+* [Endpoint](code/Endpoint.md) - responds to network requests, usually using grpc;
+* [Service](code/Service.md) - contains business logic;
+* [Database](code/Database.md) - contains peristence and data consistency logic.
+
+## Interactions between layers
+
+Peer wires together endpoints, services and databases. It does not create the database or it's own identity.
+
+Peer is organized into subsystems (e.g. overlay, kademlia).
+
+Endpoint can depend on services and databases, however it must not depend on endpoints.
+
+Service can depend on other services and databases, however it must not depend on endpoints.
+
+Databases must not depend on calling services nor endpoints.
+
+## Peer Lifecycle
+
+In principle we can think of the full flow of Storage Node and Satellite as:
+
+1. Identity of the particular peer class is loaded. This is use to uniquely identify the peer on the network.
+2. A connection(s) to the database is made.
+3. Peer is created:
+3.1. Creation takes identity and database as an argument
```

Comment:
the formatting is messed up, these `3.x` lines need a space between them to be displayed right.